File: py/zbc/message.py
# SPDX-License-Identifier: MIT
# Copyright (c) 2024-2026 ZooBC Foundation and Roberto Capodieci
"""ZBC-MSG-v1 message signing (spec/signing.md section 5)."""
import hashlib
from typing import Optional

from . import ed25519
from .address import decode_zbc_address, is_hex
from .keys import key_pair

SCHEME = "ZBC-MSG-v1"
_TAG = b"ZBC-MSG"


def message_digest(message: bytes) -> bytes:
    return hashlib.sha3_256(_TAG + message).digest()


def sign_message(seed, message: bytes) -> dict:
    kp = key_pair(seed)
    digest = message_digest(message)
    return {"scheme": SCHEME, "address": kp.address, "public_key": kp.public_key.hex(), "message_hex": message.hex(),
            "digest": digest.hex(), "signature": ed25519.sign(digest, kp.seed).hex()}


def public_key_of_address(address: str) -> Optional[bytes]:
    """The public key behind a ZBC_ address or 64 hex, or None."""
    if is_hex(address, 64):
        return bytes.fromhex(address)
    d = decode_zbc_address(address)
    return d[1] if d is not None and d[0] == "ZBC" else None


def verify_message(address: str, message: bytes, signature) -> bool:
    """False for anything that does not verify; never raises."""
    pub = public_key_of_address(address)
    if pub is None:
        return False
    if isinstance(signature, str):
        if not is_hex(signature, 128):
            return False
        signature = bytes.fromhex(signature)
    return len(signature) == 64 and ed25519.verify(message_digest(message), signature, pub)

File: rust/zbc/src/message.rs
// SPDX-License-Identifier: MIT
// Copyright (c) 2024-2026 ZooBC Foundation and Roberto Capodieci

//! ZBC-MSG-v1 message signing (spec/signing.md section 5).

use crate::address::decode_zbc_address;
use crate::encoding::is_hex;
use crate::keys::KeyPair;
use ed25519_dalek::{Signature, Verifier, VerifyingKey};
use sha3::{Digest, Sha3_256};

pub const SCHEME: &str = "ZBC-MSG-v1";

/// SHA3-256("ZBC-MSG" || message).
pub fn message_digest(message: &[u8]) -> [u8; 32] {
    let mut h = Sha3_256::new();
    h.update(b"ZBC-MSG");
    h.update(message);
    h.finalize().into()
}

#[derive(Debug, Clone)]
pub struct SignedMessage {
    pub scheme: &'static str,
    pub address: String,
    pub public_key: String,
    pub message_hex: String,
    pub digest: String,
    pub signature: String,
}

pub fn sign_message(kp: &KeyPair, message: &[u8]) -> SignedMessage {
    let digest = message_digest(message);
    SignedMessage {
        scheme: SCHEME,
        address: kp.address(),
        public_key: hex::encode(kp.public_key),
        message_hex: hex::encode(message),
        digest: hex::encode(digest),
        signature: hex::encode(kp.sign(&digest)),
    }
}

/// The public key behind a ZBC_ address or 64 hex, or None.
pub fn public_key_of_address(address: &str) -> Option<[u8; 32]> {
    if is_hex(address, 64) {
        return hex::decode(address).ok()?.try_into().ok();
    }
    let (prefix, payload) = decode_zbc_address(address)?;
    if prefix != "ZBC" {
        return None;
    }
    payload.try_into().ok()
}

/// False for anything that does not verify; never panics.
pub fn verify_message(address: &str, message: &[u8], signature: &[u8]) -> bool {
    let Some(pub_key) = public_key_of_address(address) else { return false };
    let Ok(vk) = VerifyingKey::from_bytes(&pub_key) else { return false };
    let Ok(sig) = Signature::from_slice(signature) else { return false };
    vk.verify(&message_digest(message), &sig).is_ok()
}
